Qute docs bugs (#8775)
quarkusio/quarkus: docs/src/main/asciidoc/qute-reference.adoc

<1> The `name` will be resolved against the `item.parent`.
<2> The `description` will be also resolved against the `item.parent`.

This section might also come in handy when we'd like to avoid multiple expensive invocations:

[source,html]
----

> {#with item.callExpensiveLogicToGetTheValue(1,'foo',bazinga)}

Reviewer: Damnit, `this` is back. But at least this time it makes sense.

  {#if this is "fun"} <1>
    <h1>Yay!</h1>
  {#else}
    <h1>{this} is not fun at all!</h1>
  {/if}
{/with}
----
<1> `this` is the result of `item.callExpensiveLogicToGetTheValue(1,'foo',bazinga)`. The method is only invoked once even though the result may be used in multiple expressions.

===== Let/Set Section

> This section is an alternative to `{#with}` which allows to specify one or more named parameters:

Reviewer: No it is not: it does not change the scope: it adds new lexical bindings.

```suggestion
This section allows you to define named local variables:
```

Author: > No it is not: it does not change the scope: it adds new lexical bindings.

Not really. Under the hood it does modify the current context. But I can change the wording. No problem.

Reviewer: OK, fine, it does introduce a new scope, right? It doesn't modify the current scope in the sense that it modifies the existing scope, right?

I still think that in this context, saying something along the lines of "it defines new local variables" is the easiest thing to explain to devs, no? We just want to make it clear that it's a bounded local scope with new bindings.

[source,html]
----
{#let myParent=order.item.parent}
  <h1>{myParent.name}</h1>

>   <p>Price: {price}

Reviewer: This `price` thing confuses me: it's not coming from `order.item.parent/myParent`, right?

Author: Nope. It's a mistake. Good catch! The original version had a `price` param too. Normally, Qute would attempt to find a `price` in the parent contexts...

Reviewer: Well, it's fine that it would resolve it from the parent scope (that's what I expect), I just wanted to be sure, because its presence in this example confused me.

----
<1> `it` is the result of `item.callExpensiveLogicToGetTheValue(1,'foo',bazinga)`. The method is only invoked once even though the result may be used in multiple expressions.

===== Let/Set Section

> This section is an alternative to `{#with}` which allows to specify multiple named parameters:

Reviewer: We should still remove `#{with expr as binding}` because it does not have the semantics of `with`, and can be done with `#{let binding = expr}`.

Author: Ok. The semantics of `with` is not well-defined in Qute (or in templates in general) but I agree that it could be replaced with let/set.

Reviewer: Well, `with` to modify the current scope is fine, it's the version with `as` which is problematic.

Author: Done.

<1> `showImage` is a named parameter.
<2> `it` is a special key that is replaced with the first unnamed param of the tag.

Now if we register this template under the name `itemDetail` and if we add a `UserTagSectionHelper` to the engine:

[source,java]
----
Engine engine = Engine.builder()
                   .addSectionHelper(new UserTagSectionHelper.Factory("myItem","itemDetail.html"))
                   .build();
----

NOTE: In Quarkus, all files from the `src/main/resources/templates/tags` are registered and monitored automatically!

We can include the tag like this:

[source,html]
----
<ul>
{#for item in items}
  <li>

>   {#myItem item showImage=true /} <1>

Reviewer: Huh, why did you rename it? Does the `#item` tag namespace clash with the local variable `item`? If that's the case it's a bug.

Author: No, it could be `item` or anything else. I just wanted to have the example more readable... `{#item item}` is not good.

Reviewer: OK, pfew :)

But frankly `#{item item}` is still my better than `#{myItem item}`. Perhaps it's `#{itemDescription item}` you were aiming for? ;)

Author: > But frankly `#{item item}` is still way better than `#{myItem item}`.

Oh, really? ;-) Anyway, `#{itemDescription item}` or `{#itemDetail item}` reads definitely better. My intention was to show the link between the tag name registered above. But this is not important in quarkus where tags are registered automatically.

>   {#myItem item showImage=true /} <1>

Reviewer: This is the first place I realised how we can pass more than one parameter to a tag. Is this formally documented somewhere else in this file? If not, it should, I was sure I would need a comma.

Author: Comma was never required. It's not documented. The whole SPI needs docs but since it's still a WIP it has a lower priority. In short, a section/tag may consume any number of named/unnamed parameters and it's the responsibility of a section to interprete the params, i.e. it could be an expression ('item.name') or it could be ignored (such as `in` in the `{#for item in items}`).

We already have https://github.com/quarkusio/quarkus/issues/7326.

Reviewer: So, this is what bugs me about the regularity of those tags. Some use `in` or `as` keywords (not evaluated) while others use `=` which seems to be an operator, but I'm not clear on that either.

The semantics are not clear to me as a user. And it doesn't appear regular.

Compare this with the following syntax:

```
<ul>
#{list items:products, as:'product'}
    <li>${product}</li>
#{/list}
</ul>
```

Where you can pass exactly one unnamed parameter to a tag, or you pass them by name with the `:` operator and `,` separator. As you can see, here `products` is an expression, while `'product'` is a string. The semantics are clear.

It's less flexible, but it's much more consistent with programming language constructs. For example, I'm not sure how the current semantics of passing parameters as-is to Java tag implementation combines with the rule for named parameters: can I call `#{foo first second third=something}` and expect it to work if `foo` is a Java tag?

Using the current Qute syntax, this would mean:

```
{#for items, as='item'}
```

Now, given that we only have `as` in the problematic `with` form (which we should remove in favour of `let`/`set`), that only leaves us with a single keyword `in`, and I've observed people mostly write custom tags using taglet files (not in Java), perhaps there's not much risk in this sort of keybord behaviour proliferating to confuse users.

But at the very least, we need to document that parameters are passed separated by spaces, named parameters use `=`.
